File: JavaScript/6js 2/parse.js
/*
|--------------------------------------------------------------------------
| Make the Deadfish Swim
|--------------------------------------------------------------------------
|
*/

function parse(data) {
   let res = [];
   let val = 0;
   for (let char of data) {
      if (char === "i") val++;
      else if (char === "d") val--;
      else if (char === "s") val *= val;
      else if (char === "o") res.push(val);
   }
   return res;
}


function parse(data) {
   let value = 0;
   const result = [];
   for (let i = 0; i < data.length; i++) {
      switch (data[i]) {
         case "i":
            value++;
            break;
         case "d":
            value--;
            break;
         case "s":
            value *= value;
            break;
         case "o":
            result.push(value);
            break;
         default:
            break;
      }
   }
   return result;
}


function parse(data) {
   let res = [];
   data.split("").reduce((cur, s) => {
      if (s === "i") cur++;
      if (s === "d") cur--;
      if (s === "s") cur = Math.pow(cur, 2);
      if (s === "o") res.push(cur);
      return cur;
   }, 0);
   return res;
}


console.log(parse("iiisdoso"), [8, 64]);
console.log(parse("iiisxxxdoso"), [8, 64]);

File: Python/6py 2/parse.py
"""
|--------------------------------------------------------------------------
| Make the Deadfish Swim
|--------------------------------------------------------------------------
|
"""

def parse(data):
   res, val = [], 0
   for char in data:
      if char == "i": val += 1
      elif char == "d": val -= 1
      elif char == "s": val *= val
      elif char == "o": res.append(val)
   return res


def parse(data):
   v, r = 0, []
   for c in data:
      v, r = v + {'i': 1, 'd': -1, 's': v * (v - 1)}.get(c, 0), r + [v] * (c == 'o')
   return r


def parse(data):
   value, result = 0, []
   for instruction in data:
      match instruction:
         case "i": value += 1
         case "d": value -= 1
         case "s": value = value**2
         case "o": result.append(value)
   return result


print(parse("ooo"), [0,0,0])
print(parse("ioioio"), [1,2,3])
print(parse("idoiido"), [0,1])
print(parse("isoisoiso"), [1,4,25])
print(parse("codewars"), [0])